### PR TITLE
Add section on Iterator error handling

### LINKED ISSUE #27636: Info about rejected promises in the async iterator protocol
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

### What specific section or headline is this issue about?

The async iterator and async iterable protocols

### What information was incorrect, unhelpful, or incomplete?

I was searching for information about the behavior of Async Iterators when the promise would reject.

But I found no info at all. 

### What did you expect to see?

Information about what it means when the promise returned from the next() call rejects.

For example: if I were to write an iterator that was getting some API results page-per-page, one of the http requests could fail. In that case, the promise could reject.
Then the question for me is: if the promise rejects, is the caller allowed to retry by calling next() again?
And is the iterator then expected to retry the same thing, in this case: retry the http request to the same url?

It could also be that this protocol explicitly means that the promise MUST resolve, otherwise it is not implementing the protocol, but in that case it is up to every async iterator to specify what it will return when an error occurs, and that would be a pity since we already have the rejected promise mechanism for that.

And what happens currently if an async generator function yields a rejected promise? Will it keep or rollback its 'state' (allowing you to repeat the next call), or will it completely fail? What will the result of the following next() call be in this case?

### Do you have any supporting links, references, or citations?

Maybe this discussion is somewhat related: https://github.com/tc39/proposal-async-iterator-helpers/issues/5

### Do you have anything more you want to share?

I started thinking about this because I am writing a library called [itr8 library](https://mrft.github.io/itr8). This library produces what I call "transIterators", which are basically unary functions that transform iterators, as they will take an (sync or async) iterator as input, and return another iterator as output.

I started thinking about a retry() operator, to easily abstract a retry mechansim, which can the be applied to any problem where you'd need a retry strategy.
Of course this begged the question: "when should we retry something?" A rejected promise immediately came to mind, so I checked what I could find in the docs about what the async iterator protocol said about rejected promise, which was - to my surprise - nothing at all.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/javascript/reference/iteration_protocols`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/iteration_protocols/index.md
* Last commit: https://github.com/mdn/content/commit/c2274293475b0a5b4febf85a49c1f91bf43ebac7
* Document last modified: 2023-05-03T02:24:26.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
Fix https://github.com/mdn/content/issues/27636.

@bakkot Maybe you'll be interested in taking a look—please let me know if the content is technically accurate and is what the spec considers as good convention. Thanks!

### CODE REVIEW

File: files/en-us/web/javascript/reference/iteration_protocols/index.md [hunk 1]
```diff
@@ -234,18 +234,80 @@ for (const b of obj) {
...
+
+If the invariant is broken or the `next()` method throws an error (for async iterators, it may also return a rejected promise), the error is progated to the caller. For built-in syntaxes, the iteration in progress is aborted without retrying or cleanup (with the assumption that if the `next()` method threw the error, then it has cleaned up already). If you are manually calling `next()`, you may catch the error and retry calling `next()`, but in general you should assume the iterator is already closed.
+
+If the caller decides to exit iteration for any reason, such as when it enters an error state in its own code (for example, while handling an invalid value produced by the iterator), it should call the `return()` method on the iterator, if one exists. This allows the iterator to perform any cleanup. The `return()` method is only called for premature exits—if `next()` returns `done: true`, the `return()` method is not called, with the assumption that the iterator has already cleaned up.
```
Comment (bakkot) [inline files/en-us/web/javascript/reference/iteration_protocols/index.md:270]: This should probably clarify that you shouldn't call `.return` when the  reason you're stopping is that the iterator violated the iterator protocol. Maybe:

```suggestion
If the caller decides to exit iteration for any reason other than the errors in the previous paragraph, such as when it enters an error state in its own code (for example, while handling an invalid value produced by the iterator), it should call the `return()` method on the iterator, if one exists. This allows the iterator to perform any cleanup. The `return()` method is only called for premature exits—if `next()` returns `done: true`, the `return()` method is not called, with the assumption that the iterator has already cleaned up.
```

**Review by bakkot — COMMENTED** [review]
Looks decent at a glance! You might also be interested in [this doc I wrote on this topic](https://github.com/tc39/how-we-work/blob/iterator-protocol/iterator-protocol.md) (which I haven't gotten around to cleaning up and linking anywhere).

**Comment by Josh-Cena** [comment]
Thanks for the link—that's really helpful.

**Review by bsmth — APPROVED** [review]
I'm leaving a +1 here based on reviewer feedback so you may merge when you're happy! Thanks, both!

**Comment by Josh-Cena** [comment]
I'll also assume the current content is fine :D

**Pull request merged**